logendpoint.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2017  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "logendpoint.h"
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "mainloop.h"
+#include "util.h"
+
+#define MAX_LOG_FILE_PREFIX 999
+
+static bool _logging_start_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_start_timeout();
+}
+
+static bool _alive_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_alive_timeout();
+}
+
+void LogEndpoint::_send_msg(const mavlink_message_t *msg, int target_sysid)
+{
+    uint8_t data[MAVLINK_MAX_PACKET_LEN];
+    struct buffer buffer {
+        0, data
+    };
+
+    buffer.len = mavlink_msg_to_send_buffer(data, msg);
+    _mainloop->route_msg(&buffer, target_sysid, msg->sysid);
+
+    _stat.read.total++;
+    _stat.read.handled++;
+    _stat.read.handled_bytes += msg->len;
+    if (msg->magic == MAVLINK_STX)
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink2_header);
+    else
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink1_header);
+}
+
+int LogEndpoint::_get_prefix(const char *extension, const char *dirname, uint16_t *prefix)
+{
+    struct dirent *ent;
+    DIR *dir;
+    uint16_t u;
+
+    dir = opendir(dirname);
+    if (!dir) {
+        return -errno;
+    }
+
+    *prefix = 0;
+
+    while ((ent = readdir(dir)) != nullptr) {
+        if (sscanf(ent->d_name, "%hu-", &u) == 1) {
+            if (u >= *prefix && u < MAX_LOG_FILE_PREFIX) {
+                *prefix = ++u;
+            }
+        }
+    }
+
+    closedir(dir);
+
+    return 0;
+}
+
+int LogEndpoint::_get_file(const char *extension, char **filename_result)
+{
+    char filename[PATH_MAX] = {};
```

Comment:
You could leave allocation to the caller rather than doing a "caller must free" in this function.  Passing in a pointer/pointer-len pair....